### ISSUE TITLE
RetinaNet UserWarning when creating the model

### ISSUE THREAD

**Issue by atgc1984** [fizyr/keras-retinanet#182]
Hey folks,

I am trying to implement the model on my own datasets (with CSV generator) and the results are promising.
However, after the model was created an user warning came out as below:

**train.py:66: UserWarning: Output "concatenate_2" missing from loss dictionary. We assume this was done on purpose, and we will not be expecting any data to be passed to "concatenate_2" during training.
  optimizer=keras.optimizers.adam(lr=1e-5, clipnorm=0.001))**

This happened every time when I applied different datasets.
Is there anything I should worry about? The code runs successfully and the saved H5 files could be loaded for testing. Prediction seems OK.

Thanks in advance.

**Comment by yhenon** [comment]
There is no issue here - it's just a keras warning.

To explain what's happening here: users want post- non maximum suppression detection s at test time (to use as the final detections), but they are not trained on. Having this present at train time is useful so we can save the model and then load it at test time. But keras outputs a warning that there is an unused output at train time.

I think we can keep this open for a little while in case people have any questions.

**Comment by atgc1984** [comment]
Thanks for the explain! 
Agree that we can keep this open for a while just to avoid the confusion on such Keras warning.

**Comment by de-vri-es** [comment]
I would prefer to document this elsewhere if it is a persistent "issue". I like issues to be things that need action. Does it make sense to document this in the readme?

**Comment by atgc1984** [comment]
Yes I guess we can put it in readme. This one does not affect the model running, just some potential confusion. Thanks.

**Comment by yotamDL** [comment]
Hi, 
I have a similar problem, but I can't understand if in my case it is OK as well.
I am training and testing the model on PASCAL VOC 2010, and in both cases (train and test), I get this warning:
**UserWarning: Output "non_maximum_suppression_1" missing from loss dictionary. We assume this was done on purpose, and we will not be expecting any data to be passed to "non_maximum_suppression_1" during training. sample_weight_mode=sample_weight_mode)**

In the evaluation phase, I've got exactly 300 predictions for each image, so I believe the network uses the NMS layer.
1. Should I ignore this warning to?
2. I am using a threshold of 0.5 on the scores to filter the detections, before calculating MAP. Do you have any good ideas how to improve this process? (get lower number of more accurate predictions and stop using threshold outside the network)

**Comment by de-vri-es** [comment]
> In the evaluation phase, I've got exactly 300 predictions for each image, so I believe the network uses the NMS layer.

To me that actually sounds like NMS is not performed. The warning also points in that direction.

**Comment by hgaiser** [comment]
It means NMS is an output, but not used in loss computations. The warning
can safely be ignored.

**Comment by hgaiser** [comment]
Added a comment in https://github.com/fizyr/keras-retinanet/commit/e3f9fc6105aa3968bc3fb9be491e62525a17e593 . Closing this issue.
